instrumentation/jms/jms-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jms/JmsInstrumenterFactory.java

```diff
@@ -0,0 +1,95 @@
+import static java.util.Collections.emptyList;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanKindExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.messaging.MessageOperation;
+import io.opentelemetry.instrumentation.api.instrumenter.messaging.MessagingAttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.messaging.MessagingSpanNameExtractor;
+import io.opentelemetry.instrumentation.api.internal.PropagatorBasedSpanLinksExtractor;
+import java.util.List;
+
+public final class JmsInstrumenterFactory {
+
+  private final OpenTelemetry openTelemetry;
+  private final String instrumentationName;
+  private List<String> capturedHeaders = emptyList();
+  private boolean messagingReceiveInstrumentationEnabled = false;
+
+  public JmsInstrumenterFactory(OpenTelemetry openTelemetry, String instrumentationName) {
+    this.openTelemetry = openTelemetry;
+    this.instrumentationName = instrumentationName;
+  }
+
+  @CanIgnoreReturnValue
+  public JmsInstrumenterFactory setCapturedHeaders(List<String> capturedHeaders) {
+    this.capturedHeaders = capturedHeaders;
+    return this;
+  }
+
+  @CanIgnoreReturnValue
+  public JmsInstrumenterFactory setMessagingReceiveInstrumentationEnabled(
+      boolean messagingReceiveInstrumentationEnabled) {
+    this.messagingReceiveInstrumentationEnabled = messagingReceiveInstrumentationEnabled;
+    return this;
+  }
+
+  public Instrumenter<MessageWithDestination, Void> buildProducerInstrumenter() {
+    JmsMessageAttributesGetter getter = JmsMessageAttributesGetter.INSTANCE;
+    MessageOperation operation = MessageOperation.SEND;
+
+    return Instrumenter.<MessageWithDestination, Void>builder(
+            openTelemetry,
+            instrumentationName,
+            MessagingSpanNameExtractor.create(getter, operation))
+        .addAttributesExtractor(
+            MessagingAttributesExtractor.builder(getter, operation)
+                .setCapturedHeaders(capturedHeaders)
+                .build())
+        .buildProducerInstrumenter(MessagePropertySetter.INSTANCE);
+  }
+
+  public Instrumenter<MessageWithDestination, Void> buildConsumerInstrumenter() {
+    JmsMessageAttributesGetter getter = JmsMessageAttributesGetter.INSTANCE;
+    MessageOperation operation = MessageOperation.RECEIVE;
+
+    // MessageConsumer does not do context propagation
+    return Instrumenter.<MessageWithDestination, Void>builder(
+            openTelemetry,
+            instrumentationName,
+            MessagingSpanNameExtractor.create(getter, operation))
+        .addAttributesExtractor(
+            MessagingAttributesExtractor.builder(getter, operation)
+                .setCapturedHeaders(capturedHeaders)
+                .build())
+        .setEnabled(messagingReceiveInstrumentationEnabled)
```

Comment:
The other 2 `...build()` methods seem to be missing this `setEnabled()` call. At best, it's inconsistent.

Reply:
I think this is intentional

maybe would be clearer if we renamed the containing method to `buildReceiveInstrumenter()`?

Reply:
Yep, it's intentional.

> maybe would be clearer if we renamed the containing method to `buildReceiveInstrumenter()`?

I'll rename it to `createConsumerReceiveInstrumenter()` (same as in `KafkaInstrumenterFactory`)